tests/fx/parallelization/test_tensor_parallel.py

```diff
@@ -0,0 +1,190 @@
+import unittest
+import torch
+import torch.distributed as dist
+from typing import Type
+from functools import partial
+from transformers import (
+    PretrainedConfig,
+    PreTrainedModel,
+    LlamaConfig,
+    MistralConfig,
+    LlamaForCausalLM,
+    MistralForCausalLM,
+    set_seed,
+)
+from parameterized import parameterized
+from optimum.fx.parallelization import parallelize_backend, ParallelExecutionCtx, Config
+from optimum.fx.parallelization.utils import MetaAwareMethodsPatcher, move_model_to_device, initialize_parameter_mapping
+from dist_utils import (
+    dist_init,
+    tearDown,
+    spawn,
+    gather_at_main_process,
+    NUM_AVAILABLE_DEVICES,
+    SEED
+)
+
+
+DUMMY_MODELS_TO_TEST = (
+    (LlamaForCausalLM, LlamaConfig(), ),
+    (MistralForCausalLM, MistralConfig(), ),
+)
+
+
+def dummify(config: PretrainedConfig):
+    config.num_hidden_layers = 2
+    config.use_cache = False
+    config.output_attentions = False
+    config.output_hidden_states = False
+
+def run_test_all_rank_results_match(rank: int, world_size: int, model_cls: Type[PreTrainedModel], model_config: PretrainedConfig):
+    dummify(model_config)
+
+    # initialize default group
+    dist_init(rank, world_size)
+    tp_group = dist.new_group()
+    
+    # prepare config and context
+    device = torch.device(type='cuda', index=torch.cuda.current_device())
+    ctx, cfg = ParallelExecutionCtx(tp_group=tp_group, current_device=device), Config()
+
+    inputs = {
+        "input_ids": torch.randint(low=1, high=model_config.vocab_size, size=(1, 10), device=device),
+        "attention_mask": torch.ones((1, 10), dtype=torch.int64, device=device),
+        "position_ids": torch.arange(0, 10, device=device).unsqueeze(0),
+    }
+
+    # this will initialize all linears on meta device
+    with MetaAwareMethodsPatcher():
+        model = model_cls(model_config)
+        model.eval()
+    # move model to current device, with linears still on meta, and intialize parameter mapping
+    move_model_to_device(model, device=device)
+    initialize_parameter_mapping(model, ctx=ctx)
+
+    model = torch.compile(model, fullgraph=True, backend=partial(parallelize_backend, ctx=ctx, config=cfg))
+    logits = model(**inputs)[0]
+    tensors = gather_at_main_process(tensor=logits, group=tp_group, rank=rank, world_size=world_size)
+
+    # check results at main worker process
+    if rank == 0:
+        assert len(tensors) == world_size
+        for i in range(1, world_size):
+            torch.testing.assert_close(tensors[i - 1].cpu(), tensors[i].cpu(), rtol=1e-4, atol=1e-4)
```

Comment:
it should probably be checked on all ranks

Reply:
check at the main process should be enough, because it gathers results from other ranks at main process and does comparison